cpp/Cipher/MGLGenerateKeyPairInstaller.cpp

```diff
@@ -29,44 +32,78 @@ namespace margelo {
 
 std::mutex m;
 
-// Current implementation only supports RSA schemes (check line config.variant =
-// ) As more encryption schemes are added this will require an abstraction that
-// supports more schemes
 FieldDefinition getGenerateKeyPairFieldDefinition(
     std::shared_ptr<react::CallInvoker> jsCallInvoker,
     std::shared_ptr<DispatchQueue::dispatch_queue> workerQueue) {
   return buildPair(
       "generateKeyPair", JSIF([=]) {
-        auto config = std::make_shared<RsaKeyPairGenConfig>(
+
+        KeyVariant variant =
+          static_cast<KeyVariant>((int)arguments[0].asNumber());
+        std::shared_ptr<RsaKeyPairGenConfig> rsaConfig;
+        std::shared_ptr<EcKeyPairGenConfig> ecConfig;
+
+        // switch on variant to get proper config from arguments
+        // outside of lambda 🤮
+        if (variant == kvRSA_SSA_PKCS1_v1_5 ||
+            variant == kvRSA_PSS ||
+            variant == kvRSA_OAEP
+        ) {
+          rsaConfig = std::make_shared<RsaKeyPairGenConfig>(
             prepareRsaKeyGenConfig(runtime, arguments));
+        } else
+        if (variant == kvEC) {
+          ecConfig = std::make_shared<EcKeyPairGenConfig>(
+            prepareEcKeyGenConfig(runtime, arguments));
+        } else {
+          throw std::runtime_error("KeyVariant not implemented"
+            + std::to_string((int)variant));
+        }
+
         auto promiseConstructor =
             runtime.global().getPropertyAsFunction(runtime, "Promise");
 
         auto promise = promiseConstructor.callAsConstructor(
             runtime,
             jsi::Function::createFromHostFunction(
-                runtime, jsi::PropNameID::forAscii(runtime, "executor"), 2,
-                [&jsCallInvoker, config](
+                runtime,
+                jsi::PropNameID::forAscii(runtime, "executor"),
+                4,
+                [&jsCallInvoker, variant, rsaConfig, ecConfig](
                     jsi::Runtime &runtime, const jsi::Value &,
                     const jsi::Value *promiseArgs, size_t) -> jsi::Value {
                   auto resolve =
                       std::make_shared<jsi::Value>(runtime, promiseArgs[0]);
                   auto reject =
                       std::make_shared<jsi::Value>(runtime, promiseArgs[1]);
 
-                  std::thread t([&runtime, resolve, reject,
-                                 jsCallInvoker, config]() {
+                  std::thread t([&runtime, resolve, reject, jsCallInvoker,
+                      variant, rsaConfig, ecConfig]() {
                     m.lock();
                     try {
-                      jsCallInvoker->invokeAsync([&runtime, config, resolve]() {
-                        auto keys = generateRSAKeyPair(runtime, config);
-                        auto publicKey = toJSI(runtime, keys.first);
-                        auto privateKey = toJSI(runtime, keys.second);
+                      jsCallInvoker->invokeAsync([&runtime, resolve,
```

Comment:
Wait what? What's the point of this Thread? It will immediately jump back to the JS Thread because of callInvoker->invokeAsync, why do we do that??

Nothing is running on a separate Thread here.

We probably want to call `generateKeyPair(...)` on the Thread, and instead of constructing `jsi::Value`s we construct an intermediate type first, then call back to JS using `jsCallInvoker->invokeAsync`, and then create the `jsi::Value`s and resolve the promise.